packages/base-controller/src/BaseControllerV2.test.ts

```diff
@@ -109,106 +143,258 @@ describe('BaseController', () => {
     new CountController({
       messenger: getCountMessenger(controllerMessenger),
       name: countControllerName,
-      state: { count: 0 },
+      state: {
+        people: {
+          count: 0,
+        },
+        animals: {
+          count: 0,
+          all: [],
+        },
+      },
       metadata: countControllerStateMetadata,
     });
 
     expect(controllerMessenger.call('CountController:getState')).toStrictEqual({
-      count: 0,
+      people: { count: 0 },
+      animals: {
+        count: 0,
+        all: [],
+      },
     });
   });
 
   it('should set initial schema', () => {
     const controller = new CountController({
       messenger: getCountMessenger(),
       name: 'CountController',
-      state: { count: 0 },
+      state: {
+        people: {
+          count: 0,
+        },
+        animals: {
+          count: 0,
+          all: [],
+        },
+      },
       metadata: countControllerStateMetadata,
     });
 
     expect(controller.metadata).toStrictEqual(countControllerStateMetadata);
   });
 
-  it('should not allow mutating state directly', () => {
+  it('should not allow reassigning the `state` property', () => {
     const controller = new CountController({
       messenger: getCountMessenger(),
       name: 'CountController',
-      state: { count: 0 },
+      state: {
+        people: {
+          count: 0,
+        },
+        animals: {
+          count: 0,
+          all: [],
+        },
+      },
       metadata: countControllerStateMetadata,
     });
 
     expect(() => {
-      controller.state = { count: 1 };
+      controller.state = {
+        people: {
+          count: 1,
+        },
+        animals: {
+          count: 1,
+          all: [{ name: 'giraffe' }],
+        },
+      };
     }).toThrow(
       "Controller state cannot be directly mutated; use 'update' method instead.",
     );
   });
 
+  it('should not allow reassigning an object property deep in state', () => {
```

Comment:
Nit: Have you considered adding a new mock controller for this test specifically, rather than adding all this extra state to the `CountController`? I had tried to keep the `CountController` as simple as possible to keep these tests short and easy to read.

Reply:
There's a fair bit of boilerplate in setting up a new controller, so I didn't want to crowd the top of this file with more stuff. But you're right, it was simple before, so I can revert these changes and make a new controller.

Reply:
Done in [64dfac1](https://github.com/MetaMask/core/pull/4011/commits/64dfac1d0fea1d2fd9e4384d86fe761604478bc1).